Move functionality automatically setting roughness/metallic on texture assignment to editor callback

SaracenOne opened the pull request:
This PR changes the hardcoded behaviour in BaseTexture3D where if you set assign a roughness of metallic texture to the material when one hasn't already been set, it will also set the corresponding roughness/metallic factor to 1.0 as a convenience feature to the user. However, setting it here introduces several problems. Due to the order of the properties, any time a material is loaded, since the roughness/metallic texture is loaded after its respective factor, any changes to this factor will be lost and will permenantly stuck as 1.0 regardless as what it's set at. This can introduce problems for things such as the GLTF importer, where Blender will sometimes create ORM maps for materials without metal, and make the ORM map's metallic channel white, but the corresponding metallic factor as 0.0. This gets ignored due to this bug and results in non-metallic objects becoming metallic.

In order to keep the same functionality while fixing bugs in the material loader, the code to automatically set the roughness/metallic factor to 1.0 if a new texture is added, has been moved to an UndoRedo inspector callback, so it will only change the factors if textures are explicitly assigned by the user via the inspector.

This PR must also be merged to avoid a bug which is caused by queing up both a texture and generic uniform update together (#56563).

SaracenOne committed:
Move metallic and roughness automated assignment to an inspector UndoRedo
callback.

Calinou commented:
For reference, this behavior was originally added in https://github.com/godotengine/godot/pull/54910.

SaracenOne commented:
Ah, I see. I should probably remove that part of documentation too then. I think the functionality of automatically setting the metallic and roughness factor is fine, but doing it the way you did introduced several bugs, so I think moving it to this approach makes sense.

SaracenOne commented:
Oh yeah, and one other thing, doing it this way also makes the process of using undo also reset the metallic and roughness values correctly.

akien-mga commented:
It would make sense to squash the commits, at least the 2nd and 3rd ones which are linked (the 1st one could be kept separate if it fixes a different issue).

SaracenOne commented:
Okay, I think I'll move the renderer fix to another PR since while it is required for this PR, it is fixing a different class of bug

SaracenOne commented:
Okay @akien-mga, I've squashed the commits and moved the bugfix for the RD renderer into its own seperate PR (#56563), but made it clear that both PRs will require merging for this to be fully fixed.

akien-mga commented:
Thanks!
